File: ros2_project/src/2-5custom_ros_interfaces/use_custom_msg_interfaces_cpp/src/test_custom_msg.cpp
#include <rclcpp/logging.hpp>
#include <string>
#include <rclcpp/rclcpp.hpp>
#include <rclcpp/node.hpp>
#include <custom_msg_interfaces/msg/test.hpp>


class TestCustomMsgNode : public rclcpp::Node
{
public:
    TestCustomMsgNode(std::string name):rclcpp::Node(name)
    {
        // 构建一个自定义消息类型的变量
        custom_msg_interfaces::msg::Test test_msg;
        test_msg.header.frame_id = "base_link";
        test_msg.header.stamp = this->get_clock()->now();
        test_msg.flag = true;

        // DEBUG自定义消息
        RCLCPP_INFO(this->get_logger(),"测试自定义消息：\ntest_msg.header.frame_id: %s\ntest_msg.header.stamp: %d\ntest_msg.flag: %d",test_msg.header.frame_id.c_str(),test_msg.header.stamp.sec,test_msg.flag);
    }
};

int main(int argc, char ** argv){
    rclcpp::init(argc, argv);
    auto node = std::make_shared<TestCustomMsgNode>("TestCustomMsgNode");
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: ros2_project/src/2-5custom_ros_interfaces/use_custom_msg_interfaces_py/use_custom_msg_interfaces_py/test_custom_msg.py
import rclpy
from rclpy.node import Node
from custom_msg_interfaces.msg import Test


class TestCustomMsgNode(Node):
    # 构造函数，在每个类对象初始化时都会执行
    def __init__(self,name):
        super().__init__(name)

        # 构建一个自定义消息类型的变量
        msg = Test()
        msg.header.frame_id="base_link"
        msg.header.stamp=self.get_clock().now().to_msg()
        msg.flag=True
        
        # 日志输出，debug自定义消息
        self.get_logger().info(f"测试自定义消息：\ntest_msg.header.frame_id: {msg.header.frame_id}\ntest_msg.header.stamp: {msg.header.stamp}\ntest_msg.flag: {msg.flag}")

def main(args=None):
    rclpy.init(args=args)                           # 初始化rclpy
    node = TestCustomMsgNode("TestCustomMsgNode")   # 新建一个节点
    rclpy.spin(node)                                # 保持节点运行，检测是否收到退出指令(Ctrl+C)
    rclpy.shutdown()                                # 当接收到退出指令时，退出spin函数，执行shutdown函数来关闭rclpy
